client/lib/push-notifications/index.js

```diff
@@ -0,0 +1,141 @@
+/**
+ * External dependencies
+ */
+var Emitter = require( 'lib/mixins/emitter' ),
+	debug = require( 'debug' )( 'calypso:push-notifications' );
+
+var initializeState,
+	registerServiceWorker;
+
+initializeState = function() {
+	// Only continue if the service worker supports notifications
+	if ( ! ( ( 'ServiceWorkerRegistration' in window ) && ( 'showNotification' in window.ServiceWorkerRegistration.prototype ) ) ) {
+		debug( 'Notifications not supported' );
+		return;
+	}
+
+	// If the user denied notifications permission, we cannot prompt again; the user must change the permission
+	if ( ( ! ( 'Notification' in window ) ) || 'denied' === window.Notification.permission ) {
+		debug( 'Permission denied' );
+		this.setState( 'denied' );
+		return;
+	}
+
+	// Only continue if push messaging is supported
+	if ( ! ( 'PushManager' in window ) ) {
+		debug( 'Push messaging not supported' );
+		return;
+	}
+
+	window.navigator.serviceWorker.ready.then( registerServiceWorker.bind( this ) );
+};
+
+registerServiceWorker = function( serviceWorkerRegistration ) {
+	debug( 'Registering service worker' );
+	// Grab existing subscription if we have it
+	serviceWorkerRegistration.pushManager.getSubscription().then(
+		( function( subscription ) {
```

Comment:
An arrow function would increase readability here & alleviate the need to `bind`